server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java

```diff
@@ -861,6 +862,7 @@ protected void abandonSegment(final long truncatedTime, final Sink sink)
         );
         for (FireHydrant hydrant : sink) {
           cache.close(SinkQuerySegmentWalker.makeHydrantCacheIdentifier(hydrant));
+          hydrant.getSegment().close();
```

Comment:
Despite it's documentation, `abandonSegment()` is called *not* only from `mergeExecutor`, so races between `persistAndMerge()` and `abandonSegment()` are possible. It should be resolved before merging this change, because it may lead to JVM crashes.